Taro next 找不到多端文件

<!--
  如果是提交 bug，请搜索文档和 issue，确认以下事项：
  * 该问题没有在其他 issue 和文档讨论到，不属于重复内容
  * 除了「 补充信息」外，每一都必填
  不满足以上两点要求的 bug 报告，issue 会被直接关掉
  请多多理解，您现在的不便将会使 Taro 开发者更高效地定位你的问题，修复你的问题
  像你一样的 Taro 的使用者也可以通过搜索找到你提供的 bug，对各方都有很大好处
-->

### 问题描述

@/utils/ 有两个多端文件： url.h5.js 和url.weapp.js
页面中import xxx from '@/utils/url' 报错：

```JavaScript
ModuleNotFoundError: Module not found: Error: Can't resolve '@/utils/url' ...
```

### 复现步骤

<!-- 复现问题的步骤。代码只贴截图，不贴文字会被视为无效issue -->

1. 任意文件夹下创建 url.h5.js 和url.weapp.js。

2. url.h5.js 和url.weapp.js写入以下内容

```JavaScript
export const testFn = () => {
    console.log("anything")
}
```

3. 任意js文件中import {testFn} from '@utils/url'

4. 启动和编译时报错：

```JavaScript
ModuleNotFoundError: Module not found: Error: Can't resolve '@/utils/url' ...
```

### 期望行为

启动和编译时可以顺利找到多端文件，并在下个版本中修复

### 报错信息

```JavaScript
ModuleNotFoundError: Module not found: Error: Can't resolve '@/utils/url' ...
```

### 系统信息

```shell
$ taro info
👽 Taro v3.0.0-beta.5


  Taro CLI 3.0.0-beta.5 environment info:
    System:
      OS: macOS 10.15.4
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 14.1.0 - /usr/local/bin/node
      Yarn: 1.22.4 - /usr/local/bin/yarn
      npm: 6.14.4 - /usr/local/bin/npm
    npmPackages:
      @tarojs/cli: ^3.0.0-beta.5 => 3.0.0-experimental.5 
      @tarojs/components: ^3.0.0-beta.5 => 3.0.0-beta.5 
      @tarojs/plugin-sass: 2.2.3 => 2.2.3 
      @tarojs/redux: 2.2.3 => 2.2.3 
      @tarojs/redux-h5: 2.2.3 => 2.2.3 
      @tarojs/taro: ^3.0.0-beta.5 => 3.0.0-beta.5 
      @tarojs/webpack-runner: ^3.0.0-beta.5 => 3.0.0-beta.5 
      eslint-config-taro: ^3.0.0-beta.5 => 3.0.0-experimental.5 
      react: 16.13.1 => 16.13.1 
    npmGlobalPackages:
      typescript: 3.7.4

```

### 补充信息
@yuche
在交流群中回复是 “有一些历史代码迁移没完成” 所致。

--- a/packages/taro-mini-runner/src/webpack/base.conf.ts
+++ b/packages/taro-mini-runner/src/webpack/base.conf.ts
@@ -1,5 +1,6 @@
 import * as path from 'path'
 import * as Chain from 'webpack-chain'
+import { MultiPlatformPlugin } from '@tarojs/runner-utils'
 
 export default (appPath: string) => {
   const chain = new Chain()
@@ -23,6 +24,10 @@ export default (appPath: string) => {
       ]
     }
   })
+
+  chain.resolve
+    .plugin('MultiPlatformPlugin')
+    .use(MultiPlatformPlugin, ['described-resolve', 'resolve'])
 
   return chain
 }

--- a/packages/taro-runner-utils/src/index.ts
+++ b/packages/taro-runner-utils/src/index.ts
@@ -1,7 +1,9 @@
 import { getBundleResult, getBundleContent, getSassLoaderOption } from './scss'
+import { MultiPlatformPlugin } from './resolve/MultiPlatformPlugin'
 
 export {
   getBundleResult,
   getBundleContent,
-  getSassLoaderOption
+  getSassLoaderOption,
+  MultiPlatformPlugin
 }

--- a/packages/taro-runner-utils/src/resolve/MultiPlatformPlugin.ts
+++ b/packages/taro-runner-utils/src/resolve/MultiPlatformPlugin.ts
@@ -1,0 +1,83 @@
+import * as path from 'path'
+import helper from '@tarojs/helper'
+
+interface IOptions {
+  include?: string[]
+}
+
+/**
+ * @description 此 enhance-resolve 插件用于根据当前编译的平台，解析多端文件的后缀
+ *
+ * @property {string} source resolver hook 类别
+ * @property {string} target 解析完成后需要触发的钩子
+ * @property {IOptions} options 插件配置项
+ *
+ * @example
+ *   there's filepath 'src/index'
+ *   when platform is weapp, we get 'src/index.weapp.[js|ts|jsx|tsx]'
+ *   when platform is h5, we get 'src/index.h5.[js|ts|jsx|tsx]'
+ *   by default, we get 'src/index.[js|ts|jsx|tsx]'
+ *
+ * @class MultiPlatformPlugin
+ */
+export class MultiPlatformPlugin {
+  private source: string
+  private target: string
+  private options: IOptions
+
+  constructor (source: string, target: string, options?: IOptions) {
+    this.source = source
+    this.target = target
+    this.options = options || {}
+  }
+
+  public apply (resolver) {
+    const target = resolver.ensureHook(this.target)
+    resolver
+      .getHook(this.source)
+      .tapAsync('MultiPlatformPlugin', (request, resolveContext, callback) => {
+        const innerRequest: string = request.request || request.path
+        if (!innerRequest) return callback()
+
+        if (!path.extname(innerRequest)) {
+          let srcRequest: string
+          if (path.isAbsolute(innerRequest)) {
+            // absolute path
+            srcRequest = innerRequest
+          } else if (!path.isAbsolute(innerRequest) && /^\./.test(innerRequest)) {
+            // relative path
+            srcRequest = path.resolve(request.path, request.request)
+          } else {
+            return callback()
+          }
+
+          if (/node_modules/.test(srcRequest) && !this.includes(srcRequest)) {
+            return callback()
+          }
+
+          const newRequestStr = helper.resolveMainFilePath(srcRequest)
+          if (newRequestStr === innerRequest) return callback()
+          const obj = Object.assign({}, request, {
+            request: newRequestStr
+          })
+          return resolver.doResolve(target, obj, 'resolve multi platform file path', resolveContext, (err, result) => {
+            if (err) return callback(err)
+
+            if (result === undefined) return callback(null, null)
+            return callback(null, result)
+          })
+        }
+
+        callback()
+      })
+  }
+
+  private includes (filePath: string): boolean {
+    if (!this.options.include || !this.options.include.length) return false
+
+    filePath = filePath.replace(path.sep, '/')
+
+    const res = this.options.include.find(item => filePath.includes(item))
+    return Boolean(res)
+  }
+}

--- a/packages/taro-runner-utils/types/index.d.ts
+++ b/packages/taro-runner-utils/types/index.d.ts
@@ -1,2 +1,3 @@
 import { getBundleResult, getBundleContent, getSassLoaderOption } from './scss';
-export { getBundleResult, getBundleContent, getSassLoaderOption };
+import { MultiPlatformPlugin } from './resolve/MultiPlatformPlugin';
+export { getBundleResult, getBundleContent, getSassLoaderOption, MultiPlatformPlugin };

--- a/packages/taro-runner-utils/types/resolve/MultiPlatformPlugin.d.ts
+++ b/packages/taro-runner-utils/types/resolve/MultiPlatformPlugin.d.ts
@@ -1,0 +1,27 @@
+interface IOptions {
+    include?: string[];
+}
+/**
+ * @description 此 enhance-resolve 插件用于根据当前编译的平台，解析多端文件的后缀
+ *
+ * @property {string} source resolver hook 类别
+ * @property {string} target 解析完成后需要触发的钩子
+ * @property {IOptions} options 插件配置项
+ *
+ * @example
+ *   there's filepath 'src/index'
+ *   when platform is weapp, we get 'src/index.weapp.[js|ts|jsx|tsx]'
+ *   when platform is h5, we get 'src/index.h5.[js|ts|jsx|tsx]'
+ *   by default, we get 'src/index.[js|ts|jsx|tsx]'
+ *
+ * @class MultiPlatformPlugin
+ */
+export declare class MultiPlatformPlugin {
+    private source;
+    private target;
+    private options;
+    constructor(source: string, target: string, options?: IOptions);
+    apply(resolver: any): void;
+    private includes;
+}
+export {};

--- a/packages/taro-webpack-runner/src/config/base.conf.ts
+++ b/packages/taro-webpack-runner/src/config/base.conf.ts
@@ -1,5 +1,6 @@
 import * as path from 'path'
 import * as Chain from 'webpack-chain'
+import { MultiPlatformPlugin } from '@tarojs/runner-utils'
 
 import { getRootPath } from '../util'
 import { BuildConfig } from '../util/types'
@@ -27,6 +28,10 @@ export default (appPath: string, config: Partial<BuildConfig>) => {
       modules: [path.join(getRootPath(), 'node_modules'), 'node_modules']
     }
   })
+
+  chain.resolve
+    .plugin('MultiPlatformPlugin')
+    .use(MultiPlatformPlugin, ['described-resolve', 'resolve'])
 
   return chain
 }

--- a/packages/taro-webpack-runner/src/index.ts
+++ b/packages/taro-webpack-runner/src/index.ts
@@ -72,7 +72,6 @@ const buildDev = async (appPath: string, config: BuildConfig): Promise<any> => {
   const customDevServerOption = config.devServer || {}
   const webpackChain = devConf(appPath, config)
   const onBuildFinish = config.onBuildFinish
-  console.log(config)
   await customizeChain(webpackChain, config.modifyWebpackChain, config.webpackChain)
 
   const devServerOptions = recursiveMerge<WebpackDevServer.Configuration>(

